storage/remote/queue_manager.go

```diff
@@ -302,11 +341,96 @@ func NewQueueManager(reg prometheus.Registerer, metrics *queueManagerMetrics, lo
 	}
 
 	t.watcher = wal.NewWatcher(reg, wal.NewWatcherMetrics(reg), logger, client.Name(), t, walDir)
+	if t.cfg.SendMetadata {
+		t.metadataWatcher = NewMetadataWatcher(logger, sm, client.Name(), t, t.cfg.MetadataSendInterval, flushDeadline)
+	}
 	t.shards = t.newShards()
 
 	return t
 }
 
+// AppendMetadata queues metadata to be sent to the remote storage. Metadata is sent all at once and is not parallelized.
+func (t *QueueManager) AppendMetadata(ctx context.Context, metadata []scrape.MetricMetadata) {
+	metadataSet := map[scrape.MetricMetadata]struct{}{}
+
+	// We create a set of the metadata by deduplicating based on the attributes of a
+	// scrape.MetricMetadata. In this case, a combination of metric name, help, type, and unit.
+	for _, entry := range metadata {
+		metadataSet[entry] = struct{}{}
+	}
+
+	mm := make([]prompb.MetricMetadata, 0, len(metadataSet))
+	for entry := range metadataSet {
+		mm = append(mm, prompb.MetricMetadata{
+			MetricName: entry.Metric,
+			Help:       entry.Help,
+			Type:       metricTypeToMetricTypeProto(entry.Type),
+			Unit:       entry.Unit,
+		})
+	}
+
+	err := t.sendMetadataWithBackoff(ctx, mm)
```

Comment:
No metric for failure?

also `if err := ... ; err != nil {` pattern is available here.

Reply:
Given that metadata is a best-effort approach (it doesn't even survive restarts) and the current semantics we have (sending it all at once) I don't think that's a metric we need. You can observe bytes or metadata sent to get an idea of whenever we're failing to deliver metadata or not.

Reply:
Good shout on the pattern though!

Reply:
I think in general we should match the same metrics that we have for samples.

Reply:
I have added the other two that make sense, `dropped` and `retried`.